src/acvp_hash.c

```diff
@@ -183,6 +199,226 @@ static ACVP_RESULT acvp_hash_mct_tc(ACVP_CTX *ctx,
+                                       unsigned int min_xof_bits,
+                                       unsigned int max_xof_bits) {
+    int i = 0, j = 0;
+    ACVP_RESULT rv = 0;
+    JSON_Value *r_tval = NULL;  /* Response testval */
+    JSON_Object *r_tobj = NULL; /* Response testobj */
+    unsigned int xof_len = 0;
+    unsigned int leftmost_bytes = 16;
+    unsigned int min_xof_bytes = (min_xof_bits / 8);
+    unsigned int max_xof_bytes = (max_xof_bits / 8);
+    unsigned int range = max_xof_bytes - min_xof_bytes + 1;
+
+    /*
+     * Initial Outputlen = (floor(maxoutlen/8) )*8
+     */
+    xof_len = (max_xof_bits / 8) * 8;
+    /* Convert from bits to bytes */
+    stc->xof_len = (xof_len + 7) / 8;
+
+#if 0
+    if (cap->cipher == ACVP_HASH_SHAKE_128) leftmost_bytes = 16;
+    else if (cap->cipher == ACVP_HASH_SHAKE_256) leftmost_bytes = 32;
+#endif
+
+    /* ***********
+     * OUTER LOOP
+     * ***********
+     */
+    for (j = 0; j < ACVP_HASH_MCT_OUTER; j++) {
+        /*
+         * Create a new test case in the response
+         */
+        r_tval = json_value_init_object();
+        r_tobj = json_value_get_object(r_tval);
+
+        /* ***********
+         * INNER LOOP
+         * ***********
+         */
+        for (i = 0; i <= ACVP_HASH_MCT_INNER; i++) {
+            uint16_t rightmost_out_bits = 0;
+
+            if (i != 0) {
+                /*
+                 * Use the MD[i-1] as the new Msg
+                 * Zeroize the msg buffer, and copy in the md.
+                 */
+                memzero_s(stc->msg, ACVP_HASH_MSG_BYTE_MAX);
+                if (stc->md_len <= leftmost_bytes) {
+                    memcpy_s(stc->msg, ACVP_HASH_MSG_BYTE_MAX, stc->md, stc->md_len);
+                } else {
+                    /* Only copy the leftmost 128 bits */
+                    memcpy_s(stc->msg, ACVP_HASH_MSG_BYTE_MAX, stc->md, leftmost_bytes);
+                }
+
+                if (i == ACVP_HASH_MCT_INNER) {
+                    /*
+                     * We will use the final MD as the starting MSG
+                     * for next outer loop. Break here before
+                     * doing another digest.
+                     */
+                    break;
+                }
+            }
+            stc->msg_len = leftmost_bytes;
+
+            /* Now clear the md buffer */
+            memzero_s(stc->md, ACVP_HASH_XOF_MD_BYTE_MAX);
+
+            /* Process the current SHA test vector... */
+            rv = (cap->crypto_handler)(tc);
+            if (rv != ACVP_SUCCESS) {
+                ACVP_LOG_ERR("crypto module failed the operation");
+                rv = ACVP_CRYPTO_MODULE_FAIL;
+                goto end;
+            }
+
+            /* Get the right-most 16bits and convert to an integer */
+#ifdef WIN32
+            rightmost_out_bits = _byteswap_ushort(*(uint16_t *)(stc->md + stc->md_len - 2));
+#else
+# if ACVP_HOST_LITTLE_ENDIAN
```

Comment:
Got to be a 16 bit byte swap libc API that is endian agnostic I would think. Maybe not that supports windows and Mac ?

Reply:
Do you have any specific in mind? I think ntohs is also endian dependent